cpp/src/buffer/buffer.cpp

```diff
@@ -139,8 +152,148 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
+                });
+            }
+        },
+        storage_
+    );
+}
+
+std::unique_ptr<Buffer> Buffer::copy_slice(
+    MemoryType target,
+    std::ptrdiff_t offset,
+    std::ptrdiff_t length,
+    rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(offset <= std::ptrdiff_t(size), "offset can't be more than size");
+    RAPIDSMPF_EXPECTS(
+        offset + length <= std::ptrdiff_t(size), "offset + length can't be more than size"
+    );
+
+    if (mem_type() == target) {
+        return copy_slice(offset, length, stream);
+    }
+
+    // Implement the copy between each possible memory types (both directions).
+    return std::visit(
+        overloaded{
+            [&](const HostStorageT& storage) {  // host -> device
+                return std::unique_ptr<Buffer>(new Buffer{
+                    std::make_unique<rmm::device_buffer>(
+                        static_cast<uint8_t const*>(storage->data()) + offset,
+                        length,
+                        stream,
+                        br->device_mr()
+                    ),
+                    stream,
+                    br
+                });
+            },
+            [&](DeviceStorageT const& storage) {  // device -> host
+                {
+                    auto ret = std::make_unique<std::vector<uint8_t>>(length);
+                    RAPIDSMPF_CUDA_TRY_ALLOC(cudaMemcpyAsync(
+                        ret->data(),
+                        static_cast<cuda::std::byte const*>(storage->data()) + offset,
+                        size_t(length),
+                        cudaMemcpyDeviceToHost,
+                        stream
+                    ));
+                    auto host_buf =
+                        std::unique_ptr<Buffer>(new Buffer{std::move(ret), br});
+                    // Create a new event to track the async copy only if length > 0
+                    if (length > 0) {
+                        host_buf->override_event(std::make_shared<Event>(stream));
+                    }
+                    return host_buf;
+                }
+            }
+        },
+        storage_
+    );
+}
+
+std::ptrdiff_t Buffer::copy_to(
+    Buffer& dest, std::ptrdiff_t dest_offset, rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(
+        dest.size - size_t(dest_offset) >= size,
+        "destination buffer is too small",
+        std::invalid_argument
+    );
+
+    if (size == 0) {  // empty buffer, nothing to do
+        return 0;
+    }
+
+    cudaMemcpyKind kind;
+
+    // if both buffers are on the host, use memcpy, otherwise, use cudaMemcpyAsync
+    if (mem_type() == MemoryType::HOST && dest.mem_type() == MemoryType::HOST) {
+        // if the source buffer is not ready, return 0
+        if (!is_ready()) {
+            return 0;
+        }
+        // src buffer is ready, copy the data
+        std::memcpy(static_cast<uint8_t*>(dest.data()) + dest_offset, data(), size);
+        return std::ptrdiff_t(size);
```

Comment:
Same here, maybe we should consider `cudaMemcpyAsync(..., cudaMemcpyHostToHost, stream)`.

Reply:
Ah. yes, that would simplify things a lot.

Reply:
@pentschev after reading the docs, I feel like it is not async. 
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79

Reply:
Please check this guide 
https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html

Reply:
That seems to suggest we could just use `cudaMemcpyHostToHost` for this case too. It might be blocking but the `std::memcpy` is "blocking" too.

Reply:
@wence- My idea was, in the case of `Buffer::copy_to`, we can return `0` if the source buffer is not ready and then the caller can call `copy_to` again later on. But I also agree that using `cudaMemcpyHostToHost` is a safer option. 
Any preferences?

Reply:
I think for this PR I'd prefer we switch to `cudaMemcpyHostToHost` and try to address a more efficient pipeline where we can make all non-blocking in a follow-up PR, it's best to avoid addressing multiple issues at once.